db/python/tables/output_file.py

```diff
@@ -0,0 +1,266 @@
+                # This will set this client object to point to the local google cloud storage.
+                client_options={'api_endpoint': 'http://localhost:4443'},
+            )
+        else:
+            client = Client()
+
+        async with self.connection.transaction():
+            if 'main_files' in files:
+                for primary_file in files['main_files']:
+                    parent_file_id = await self.create_or_update_output_file(
+                        path=primary_file['basename'], client=client
+                    )
+                    await self.add_output_file_to_analysis(
+                        analysis_id,
+                        parent_file_id,
+                        json_structure=primary_file['json_path'],
+                        # If the file couldnt be created, we just pass the basename as the output
+                        output=None if parent_file_id else primary_file['basename'],
+                    )
+                    if files.get('secondary_files_grouped'):
+                        secondary_files = files['secondary_files_grouped']
+                        if primary_file['basename'] in secondary_files:
+                            for secondary_file in secondary_files[
+                                primary_file['basename']
+                            ]:
+                                secondary_file_id = (
+                                    await self.create_or_update_output_file(
+                                        path=secondary_file['basename'],
+                                        parent_id=parent_file_id,
+                                        client=client,
+                                    )
+                                )
+                                await self.add_output_file_to_analysis(
+                                    analysis_id,
+                                    secondary_file_id,
+                                    json_structure=secondary_file['json_path'],
+                                    # If the file couldnt be created, we just pass the basename as the output
+                                    output=(
+                                        None
+                                        if secondary_file_id
+                                        else secondary_file['basename']
+                                    ),
+                                )
+                                file_ids.append(secondary_file_id)
+                    file_ids.append(parent_file_id)
+
+        client.close()
+        # check that only the files in this json_dict should be in the analysis. Remove what isn't in this dict.
+        _update_query = dedent(
+            """
+            DELETE ao FROM analysis_outputs ao
+            WHERE (ao.analysis_id = :analysis_id)
+            AND (ao.file_id NOT IN :file_ids)
+            """
+        )
+
+        if file_ids:
+            await self.connection.execute(
```

Comment:
I think it'd be good if this was in the same transaction as the create/update, otherwise the updates won't be rolled back if the deletion fails